pkg/queue/sharedmain/main.go

```diff
@@ -243,52 +253,45 @@ func Main(opts ...Option) error {
 		concurrencyendpoint = queue.NewConcurrencyEndpoint(env.ConcurrencyStateEndpoint, env.ConcurrencyStateTokenPath)
 	}
 
-	// Enable TLS when certificate is mounted.
-	tlsEnabled := exists(logger, certPath) && exists(logger, keyPath)
+	var srvs []service
+	var drainerTLS *pkghandler.Drainer
 
-	mainServer, drainer := buildServer(d.Ctx, env, d.Transport, probe, stats, logger, concurrencyendpoint, false)
-	httpServers := map[string]*http.Server{
-		"main":    mainServer,
-		"metrics": buildMetricsServer(protoStatReporter),
-		"admin":   buildAdminServer(d.Ctx, logger, drainer),
-	}
-	if env.EnableProfiling {
-		httpServers["profile"] = profiling.NewServer(profiling.NewHandler(logger, true))
-	}
+	// Enable TLS when certificate is mounted.
+	encryptionEnabled := exists(logger, certPath) && exists(logger, keyPath)
 
-	// Enable TLS server when activator server certs are mounted.
 	// At this moment activator with TLS does not disable HTTP.
+	// Start main service regardless if tlsEnabled is true
 	// See also https://github.com/knative/serving/issues/12808.
-	var tlsServers map[string]*http.Server
-	if tlsEnabled {
-		mainTLSServer, drainer := buildServer(d.Ctx, env, d.Transport, probe, stats, logger, concurrencyendpoint, true /* enable TLS */)
-		tlsServers = map[string]*http.Server{
-			"tlsMain":  mainTLSServer,
-			"tlsAdmin": buildAdminServer(d.Ctx, logger, drainer),
-		}
-		// Drop admin http server as we Use TLS for the admin server.
-		// TODO: The drain created with mainServer above is lost. Unify the two drain.
-		delete(httpServers, "admin")
+	mainService, drainer := buildMainService(d.Ctx, env, d.Transport, probe, stats, logger, concurrencyendpoint, false)
+	srvs = append(srvs, mainService)
+	if encryptionEnabled {
+		var mainServiceTLS service
+		mainServiceTLS, drainerTLS = buildMainService(d.Ctx, env, d.Transport, probe, stats, logger, concurrencyendpoint, true)
```

Comment:
@dprotaso  - afaik, we use to have two unique servers that don't share http handlers and we now continue to have two unique servers that don't share http handlers. This PR has not introduced this and is not aiming to change the HTTP handlers, instead, this PR try to fix some issues identified in the issue with the initialization and termination. If there are additional problems, I suggest we open a new issue on them and try to fix them.

Reply:
> This PR has not introduced this and is not aiming to change the HTTP handlers, instead, this PR try to fix some issues identified in the issue with the initialization and termination

I'm a bit mixed because the PR is titled 'fix intialization' but it compounds on the prior problems/tech debt that this area of the code had (ie. introducing two drainers when the queue proxy should only have one).

Maybe it's worth doing the handler refactor prior to this change

Reply:
This PR relates to https://github.com/knative/serving/issues/13720  - this issue does not discuss any handler refactoring, it identified a number of other issues which this PR seek to solve.